torch/onnx/_internal/fx/dynamo_exporter.py

```diff
@@ -28,41 +40,206 @@ def export(self) -> torch.onnx.ExportOutput:
 
         class GraphCaptureCompiler:
             def __init__(self):
-                self.captured_graph: Optional["torch.fx.GraphModule"] = None
+                self.captured_graph: Optional[torch.fx.GraphModule] = None
                 self.captured_graph_count = 0
 
-            def compile(self, graph_module: "torch.fx.GraphModule", _):
+            def compile(self, graph_module: torch.fx.GraphModule, _):
                 assert self.captured_graph_count == 0
                 self.captured_graph = graph_module
                 self.captured_graph_count += 1
                 return graph_module
 
         compiler = GraphCaptureCompiler()
         torch._dynamo.reset()
-        torch._dynamo.optimize(compiler.compile, nopython=True)(self.model)(*bound_args)
+        # TODO(titaiwang): Set `dynamic` according to `self.options.dynamic_shapes`
+        torch._dynamo.optimize(compiler.compile, nopython=True)(self.model)(*model_args)
         torch._dynamo.reset()
         assert compiler.captured_graph
+
+        # Outputs are flattened by `dynamo.optimize`.
+        # Apply and record this output format step.
+        self._output_formatter.append_step(DynamoFlattenOutputStep())
+        # TODO: `dynamo.optimize` does not capture non computation part of the graph.
+        # Hence any tensor that is returned multiple times will only appear once
+        # in the return statement of `dynamo.optimize` traced graph. A specialized
+        # output formatter is required to map this gap between PyTorch model.
+
         # Export FX graph to ONNX ModelProto.
-        return self.export_fx_to_onnx(
-            compiler.captured_graph, tuple(arg for arg in bound_args if arg is not None)
+        #
+        # Apply and record the following input format steps.
+        # `args` and `kwargs` are merged and flattened by `dynamo.optimize`.
+        model_args, _ = self._apply_input_format_step(
+            fx_exporter.FlattenInputWithTreeSpecValidationStep, model_args, {}
+        )
+        # `None` inputs are removed by `dynamo.optimize`.
+        model_args, _ = self._apply_input_format_step(
+            fx_exporter.RemoveNoneInputStep, model_args, {}
+        )
+        # TODO: needs additional input format step to remove unused arguments.
+        # `dynamo.optimize` drops unused arguments.
+        return self.export_fx_to_onnx(compiler.captured_graph, model_args)
+
+
+class _PyTreeExtensionContext:
```

Comment:
can you elaborate on why we need to extend pytree?

Reply:
https://github.com/pytorch/pytorch/issues/98251

Reply:
Not for this PR, this almost feels like deserving a deeper discussion. 

If Dynamo cannot support custom types, maybe we could propose an interface for users to register any custom data type, so that exporter knows how to serialize? We can do that at dynamo or exporter context, probably? In fact, we discussed this idea in the past for the torchscript exporter.